### PR TITLE
fix torch.cat bug with boxed CPUFallback

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/60902

The boxed fallback was written to assume that there was at least one tensor argument, which it used to figure out what device to move the cpu tensors to. That fails with an op like `torch.cat()`, which doesn't have any tensor arguments, but instead has a single `TensorList` argument. It looks like that wasn't caught originally because there weren't any ops that xla tests CPU fallbacks for that have only TensorList args.

I also added handling to gracefully deal with the case where you have an empty list of tensors - in that case we don't know what device to move everything to, but that doesn't matter because an empty list of tensors implies that we have no tensors to move anyway.

I tested it out though and noticed that `torch.cat(())` doesn't handle empty lists well anyway (erroring out in the dispatcher). I'm not sure that it's a huge issue, and not even sure that we want to fix it (default to CPU? add an extra codegen'd check into every op that only takes TensorList args?) but I'll file a separate bug for that: https://github.com/pytorch/pytorch/issues/60997

I tested it by running the pytorch/xla suite after removing `cat` from `xla_native_functions.yaml`, and confirming that we don't segfault anymore.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60993 fix torch.cat bug with boxed CPUFallback**

Differential Revision: [D29471577](https://our.internmc.facebook.com/intern/diff/D29471577)

### CODE REVIEW

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29471577).

File: aten/src/ATen/native/CPUFallback.cpp [hunk 1]
```diff
@@ -135,18 +141,46 @@ void cpu_fallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
...
+          // Decide what device to move the output tensor(s) to.
+          // The current convention is that we use the first tensor arg to pick the device
+          // Barring that, we take the first tensor from a TensorList arg.
+          c10::optional<c10::Device> tgt_device = c10::nullopt;
+          if (tensor_args.size() > 0) {
+              tgt_device = tensor_args[0].device();
+          } else {
+            // We need to loop through all of the (potentially multiple) TensorList arguments
+            // In case, e.g. the first one is empty but the second is not.
+            for (auto& tens_list : tensorlist_args) {
```
Comment (asuhan) [inline aten/src/ATen/native/CPUFallback.cpp:153]: This should go in a helper, it's too hard to tell right away what these nested loops are doing.

**Review by asuhan — CHANGES_REQUESTED** [review]

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29471577).

**Review by asuhan — APPROVED** [review]

File: aten/src/ATen/native/CPUFallback.cpp [hunk 2]
```diff
@@ -39,6 +40,25 @@ std::vector<at::Tensor> to_cpu(const at::TensorList& tensors) {
   return cpu_tensors;
 }
 
+c10::optional<c10::Device> compute_target_device(std::vector<at::Tensor>& t_args, std::vector<c10::List<at::Tensor>> tlist_args) {
+  // Decide what device to move the output tensor(s) to.
+  // The current convention is that we use the first tensor arg to pick the device
+  // Barring that, we take the first tensor from a TensorList arg.
+  c10::optional<c10::Device> tgt_device = c10::nullopt;
```
Comment (asuhan) [inline aten/src/ATen/native/CPUFallback.cpp:47]: This is unused.

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29471577).

**Review by ezyang — APPROVED** [review]
okey dokey

**Pull request closed without merging**